crates/wasmparser/src/parser.rs

```diff
@@ -1,1313 +1,1338 @@
-/* Copyright 2017 Mozilla Foundation
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-// See https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md
-
-use std::boxed::Box;
-use std::vec::Vec;
-
-use crate::limits::{
-    MAX_WASM_FUNCTIONS, MAX_WASM_FUNCTION_LOCALS, MAX_WASM_STRING_SIZE, MAX_WASM_TABLE_ENTRIES,
-};
-
-use crate::primitives::{
-    BinaryReaderError, CustomSectionKind, ExternalKind, GlobalType, ImportSectionEntryType,
-    LinkingType, MemoryType, Naming, Operator, RelocType, Result, SectionCode, TableType, Type,
-    TypeDef,
-};
-
-use crate::readers::*;
-
-use crate::binary_reader::{BinaryReader, Range};
-
-const MAX_DATA_CHUNK_SIZE: usize = MAX_WASM_STRING_SIZE;
-
-#[derive(Debug)]
-pub struct LocalName<'a> {
-    pub index: u32,
-    pub locals: Box<[Naming<'a>]>,
+use crate::{AliasSectionReader, InstanceSectionReader, ModuleSectionReader};
+use crate::{BinaryReader, BinaryReaderError, FunctionBody, Range, Result};
+use crate::{DataSectionReader, ElementSectionReader, ExportSectionReader};
+use crate::{FunctionSectionReader, ImportSectionReader, TypeSectionReader};
+use crate::{GlobalSectionReader, MemorySectionReader, TableSectionReader};
+use std::convert::TryInto;
+use std::fmt;
+use std::iter;
+
+/// An incremental parser of a binary WebAssembly module.
+///
+/// This type is intended to be used to incrementally parse a WebAssembly module
+/// as bytes become available for the module. This can also be used to parse
+/// modules that are already entirely resident within memory.
+///
+/// This primary function for a parser is the [`Parser::parse`] function which
+/// will incrementally consume input. You can also use the [`Parser::parse_all`]
+/// function to parse a module that is entirely resident in memory.
+#[derive(Debug, Clone)]
+pub struct Parser {
+    state: State,
+    offset: u64,
```

Comment:
It would be nice to have a method that exposes this value (or maybe have the offset in `Chunk::Parsed` or `Payload`?). We need it for processing DWARF.

I know it can be reconstructed from looking at `Chunk::Parsed::consumed`, but (a) that seems quite tedious, and (b) might not have the granularity required for the DWARF stuff (offset per instruction).

Reply:
Nevermind: just saw `OperatorsReader::read_with_offset` etc.

Reply:
Ah yeah FWIW I also want to make sure that every single byte in a wasm file can be accounted for using this API (one way or another at least). The `wasmparser-dump` crate (which previously lived at `tests/dump.rs`) ensures this too